iio: gyro: Update ADXRS290 driver for channels support

layman-n-ish opened the pull request:
This patch adds channels support for angular rate data & temperature data and the ability to program the 3db frequencies for both low-pass & high-pass filter all through the SPI protocol. This patch also updates the dt-overlay to configure ADXRS290 to SPI mode 3 and correspondingly updates the binding document.

Signed-off-by: Nishant Malpani <nish.malpani25@gmail.com>

layman-n-ish commented:
Below are a few queries I still have, hence the "WIP". It'd be of immense help if someone could comment on those; thanks in advance!

- I'm guessing in [`adxrs290_probe()`](https://github.com/layman-n-ish/linux/blob/rpi-4.19.y_adxrs290/drivers/iio/gyro/adxrs290.c#L488), it'd be enough to check only one of DEV_ID, MEMS_ID, ADI_ID - but which one?
- The angular data rate is 16 bits, in two 8 bit registers. I, currently, read the rate data by [exercising 2 successive `spi_w8r8()` calls](https://github.com/layman-n-ish/linux/blob/rpi-4.19.y_adxrs290/drivers/iio/gyro/adxrs290.c#L130); first on the lower address register and then on the lower-address+1 register (the higher-address one). As @buha and @dbogdan have pointed out, the CS would still be toggled in between the 2 successive calls to the same chip and the data between the calls might change. But this current solution works for both *measurement* mode and *standby* mode (where the device cannot handle reading multiple bytes in one SPI command). So, how should I proceed?
- In `adxrs290_read_raw`, I [read the "cached" 3db frequency](https://github.com/layman-n-ish/linux/blob/rpi-4.19.y_adxrs290/drivers/iio/gyro/adxrs290.c#L377); I don't actually read from the register. Is that acceptable?
- The current implementation of [finding the *closest* 3db frequency](https://github.com/layman-n-ish/linux/blob/rpi-4.19.y_adxrs290/drivers/iio/gyro/adxrs290.c#L227) with the acceptable ones only considers the integer part of the frequency. Taking into account the resolution of acceptable 3db frequencies for the high-pass filter (pasted below) this ain't a good solution (the integer part is almost the same for the majority). Should I progressively compare the fractional part after the integer part or should I do a 'float' arithmetic directly (which I didn't previously because I was under the impression that floating-point arithmetic in the kernel space is frowned upon)?

```
# supported 3db frequencies of the high-pass filter
0.000 0.011 0.022 0.044 0.087 0.175 0.350 0.700 1.400 2.800 11.300
```
- I go into the *measurement* mode and switch on the temperature sensor [at the probe callback](https://github.com/layman-n-ish/linux/blob/rpi-4.19.y_adxrs290/drivers/iio/gyro/adxrs290.c#L500) every time and stay in the same state throughout. How and when should I configure to go to the *standby* (low-power) mode?

commodo commented on drivers/iio/gyro/adxrs290.c:
```diff
@@ -10,18 +10,390 @@
+	[10] = {11, 300},
+};
+
+/**
```
doc-strings aren't usually needed for these type of static functions in the driver;
they're IIO API functions, so this is documenting what should be documented in IIO;

commodo replied:
wait; sorry; this is an internal function to the driver;
even so, they are not typically documented;

layman-n-ish replied:
Okay, I thought it'd be a good-to-have detail. Will remove in v2.

commodo commented on drivers/iio/gyro/adxrs290.c:
```diff
@@ -10,18 +10,390 @@
+ * Return:	0 if successful reads, or a negative error code on failure of
+ *		either read
+ */
+static int adxrs290_spi_read16b(struct iio_dev *indio_dev, u8 cmd, int *lsb,
```
wouldn't a normal SPI-transfer call work here?
it looks like this may be re-implementing what could be done with a standard spi_read() or spi_sync_transfer() call

layman-n-ish replied:
You're right but as I mentioned in a comment above (see pt. 2), the CS line toggles between the two successive 8b reads, in the current solution of `adxrs290_spi_read16b`, which would work for both _standby_ and _measurement_ mode. The datasheet says:

> In standby mode, only single-address SPI transactions are performed, which includes reading from or writing to a single register, but does not include writing to or reading from several registers in one command.

From the available SPI API, I previously tried `spi_w8r16` (in _measurement_ mode) for multiple bytes read but that didn't work as expected. From what I see in source code of [`drivers/spi/spi.c`](https://elixir.bootlin.com/linux/latest/source/drivers/spi/spi.c#L3865), I don't think it auto-increments to read the next address from the address provided, does it? This also brings to my next question of: how can one exercise multiple bytes read with one SPI command? The datasheet doesn't explicitly mention this. SPI timing diagram from ADXRS290's datasheet:

![adxrs290_spi_timing_dia](https://user-images.githubusercontent.com/33180939/86569696-6f308980-bf8c-11ea-9660-e0b5c7760566.png)

commodo replied:
but if you use `spi_sync_transfer()` you can set a `cs_change = 1` parameter between 2 transfers;
so, you could define 2 transfers and one toggles the CS change;
it's a slightly more common approach;

layman-n-ish replied:
Yes, I've seen that in a few drivers. I'll do that then! 

Sorry; one more thing - the current `adxrs290_spi_read16b` function should still be present for me to read 16b in _standby_ mode, right? Depending on my current state (_standby_ or _measurement_), I'll have to call a corresponding read16b function - sort of like dynamic dispatch. Am I going in the right direction?

layman-n-ish replied:
Oh wait, I think I misread your previous comment. You want me to have 2 different functions with SPI transfers to read 16b - one which doesn't toggle the CS (to use in _measurement_ mode) and one which does (to use in _standby_ mode). Did I infer it correctly?

commodo replied:
> I don't think it auto-increments to read the next address from the address provided, does it? (Or does the hardware do it?) This also brings to my next question of: how can one exercise multiple bytes read with one SPI command? The datasheet doesn't explicitly mention this. SPI timing diagram from ADXRS290's datasheet:

I forgot to answer this part initially: so, the SPI framework should handle most things here. Maybe give it a try and read and see how it works.

If there some differences between measurement mode and they aren't too big, you could implement a:
```
static int adxrs290_spi_read16b(... bool standby)
{
     if (standby) {

     } else {

     }
}
```

and call it with adxrs290_spi_read16b(..., false) for measurement mode;

layman-n-ish replied:
Got it! This clears a lot of things; thanks. Will address this in v2.

commodo commented on drivers/iio/gyro/adxrs290.c:
```diff
@@ -61,6 +479,42 @@ static int adxrs290_probe(struct spi_device *spi)
+	val = spi_w8r8(spi, ADXRS290_READ | ADXRS290_REG_ADI_ID);
+	if (val != ADXRS290_ADI_ID) {
+		dev_err(&spi->dev, "Wrong ADI ID 0x%02x\n", val);
+		return -EINVAL;
```
this error looks more like -ENODEV

commodo replied:
well, all these chip error checks look like they should return -ENODEV;

layman-n-ish replied:
Oh yes! Thanks for pointing out. Will address in v2.

commodo commented on drivers/iio/gyro/adxrs290.c:
```diff
@@ -61,6 +479,42 @@ static int adxrs290_probe(struct spi_device *spi)
+	if (ret < 0)
+		return ret;
+
+	ret = adxrs290_get_3db_freq(indio_dev, ADXRS290_LPF_MASK,
```
i don't know what the correct approach would be here [for these settings];
for some settings it's the driver that should set a default value [on driver init], for other settings it's probably ok to have the default-on-reset value;

i don't know if there is a general consensus on which is the more appropriate;
maybe upstream would tell us;
but just keep this in mind, when sending this upstream;

for now, this is fine from my side;

layman-n-ish replied:
Great point! Will keep this in mind when sending it to upstream. Thanks.

commodo commented:
rest looks good

nunojsa commented on drivers/iio/gyro/adxrs290.c:
```diff
@@ -10,18 +10,390 @@
+			return IIO_VAL_INT_PLUS_NANO;
+		case IIO_TEMP:
+			*val = 100;
+			*val2 = 0;
```
You also mentioned that we have 12 LSB/deg sensitivity for temperature... Having it to 0.1°C seems like a rough approximation.

layman-n-ish replied:
Okay, will remove the part where I set `*val2` while returning `IIO_VAL_INT`. 

Yes, I've made sure the units are appropriately set as dictated in the ABI.

layman-n-ish replied:
Sorry, the sensitivity is 10 LSB/deg Celsius for temperature. There's a typo in the commit message - will fix it in v2. Thanks for pointing it out! :)

nunojsa replied:
Ok, then nothing to add. But yes, please update the commit message...

nunojsa commented on drivers/iio/gyro/adxrs290.c:
```diff
@@ -10,18 +10,390 @@
+ * Return:	0 if successfully read, or a negative error code on failed read
+ */
+static int adxrs290_get_rate_data(struct iio_dev *indio_dev, u8 cmd,
+				  unsigned int *val)
```
It looks like `cmd` can be a `const u8`. It is a good practice to define things as `const` when they are...

layman-n-ish replied:
You're right. I'll _constify_ wherever possible in v2. Thanks for the tip!

layman-n-ish commented on drivers/iio/gyro/adxrs290.c:
```diff
@@ -10,18 +10,390 @@
+	int temp;
+	u8 cmd;
+
+	cmd = ADXRS290_READ_REG(ADXRS290_REG_FILTER);
```
I'm not entirely sure if I understand what you mean here. Do you mean to replace adxrs290_get_3db_freq() with a generic adxrs290_spi_read8b() function? But we already have SPI core's available API - spi_w8r8() - to do so.

If I'm wrong in inferring what you meant, could you please elaborate your point?

buha replied:
I misunderstood the point of this function, i now checked the datasheet and i understand you are doing a very specific thing here which won't be used elsewhere so there is no point in code reuse.

buha commented on arch/arm/boot/dts/overlays/rpi-adxrs290-overlay.dts:
```diff
@@ -24,6 +24,8 @@
 				compatible = "adi,adxrs290";
 				reg = <0>;
 				spi-max-frequency = <5000000>;
+				spi-cpol;
```
I am wondering how did this previously work (cpol=0, cpha=0 I assume) without junk readings ?

layman-n-ish replied:
It didn't, haha. The skeleton driver back was too naive to interface with SPI and I hadn't read the datasheet in detail to realise this back then.

buha replied:
That makes sense !

layman-n-ish commented:
Changelog v2:

- Restructured `adxrs290_spi_read16b()` to use `spi_sync_transfer()`. This change required introducing data (tx and rx) buffers in `adxrs290_state` and also a chip mode in `adxrs290_state` to make the driver conscious of which mode (STANDBY vs MEASUREMENT) the device is in.
- _constify_ variables at the appropriate places
- removed doc-strings
- removed `*val2 = 0` when returning IIO_VAL_INT
- replace -EINVAL with -ENODEV
- sleep for 100ms after switching from _standby_ mode to _measurement_ mode.

buha commented on drivers/iio/gyro/adxrs290.c:
```diff
@@ -30,13 +354,57 @@ static int adxrs290_write_raw(struct iio_dev *indio_dev,
+	switch (mask) {
+	case IIO_CHAN_INFO_LOW_PASS_FILTER_3DB_FREQUENCY:
+		lpf_idx = adxrs290_find_best_match(adxrs290_lpf_3db_freq_tbl,
+					ARRAY_SIZE(adxrs290_lpf_3db_freq_tbl),
```
I think you might get some comments about formatting here. Typical thing that it's done is to align parameters 2, 3 ... with the first character of the first parameter like:
```
lpf_idx = adxrs290_find_best_match(adxrs290_lpf_3db_freq_tbl,
tab tab tab ... space space space  ARRAY_SIZE(adxrs290_lpf_3db_freq_tbl),
...
```

buha replied:
Btw, general question, did you run checkpatch.pl ?

layman-n-ish replied:
I previously aligned it as you pointed out but then the line exceeds the 80 characters conventional rule.

Yes, I ran `./scripts/checkpatch.pl -f drivers/iio/gyro/adxrs290.c` on this; gives me no warning on this line.

buha replied:
I think that rule was anyhow not mandatory in cases where it's impossible to apply like this one.
Also, recently, there was a belief system shift, I guess in newer kernels checkpatch stops checking this :D But again, I can't be 100% sure what the upstream will tell you...
https://linux.slashdot.org/story/20/05/31/211211/linus-torvalds-argues-against-80-column-line-length-coding-style-as-linux-kernel-deprecates-it

layman-n-ish replied:
Interesting article! I agree with the points Linus promotes too (which is why I left [this](https://github.com/layman-n-ish/linux/blob/rpi-4.19.y_adxrs290/drivers/iio/gyro/adxrs290.c#L264) line for better readability & for 'grep' to work even though it generates a warning). I'll align the parameters with the first parameter in v3.

buha commented on drivers/iio/gyro/adxrs290.c:
```diff
@@ -5,23 +5,347 @@
+	struct spi_device	*spi;
+	/* To atomize successive reads for single measurement */
+	struct mutex		lock;
+	union {
```
Why would this be a union ? On SPI you transfer on MOSI and MISO at the same time hence using both buffers at the same time so you need separate memory areas.

You also transfer the same amount of data so their .len in spi_transfer below should be equal.

I would have a tx[3] and rx[3] in the scenario where you read 2 consecutive registers (1 byte address, reg 1, reg 2).

buha replied:
Common practice is to put these buffers right at the end of the device struct and append the first buffer with `____cacheline_aligned` like you did for the union.

layman-n-ish replied:
The timing diagram threw me off to the direction of using a `union` (quite a few drivers do it too) - the data in rx can be ignored until the tx buffer has been shifted out. Only after the tx transfer has been completed, the data in the rx buffer should be "recorded", which is what I do while initialising the `struct spi_transfer`.
![adxrs290_spi_timing_dia](https://user-images.githubusercontent.com/33180939/87129272-ab2a5e00-c2ae-11ea-9fcd-06ecfd629e93.png)

layman-n-ish replied:
Sure, will add the buffers at the end of `struct adxrs290_state` in v3.

layman-n-ish replied:
Example of a driver which uses a `union` in a similar way: [click here](https://github.com/analogdevicesinc/linux/blob/rpi-4.14.y_GSOC_2019/drivers/iio/adc/ad7292.c).

buha replied:
as discussed offline, this looks ok in your particular situation, marking as resolved

layman-n-ish commented:
Changelog v3:

- move SPI tx, rx data buffers at the end of device struct (`adxrs290_state`)
- align parameters with the first character of the first parameter even if the characters limit per line (80 characters according to checkpatch.pl) is violated
- removed conditional for `cs_toggle` which checks the mode (_standby_ vs _measurement_) of the device before enabling/disabling toggling of CS between transfers, because 16b SPI read of rate data is valid only in _measurement_ mode (see [Datasheet](https://www.analog.com/media/en/technical-documentation/data-sheets/ADXRS290.pdf) pg. 11) and ergo it doesn't make sense to add support for reading it in _standby_ mode.

(Thanks for the review!)

buha commented on drivers/iio/gyro/adxrs290.c:
```diff
@@ -5,23 +5,344 @@
+	if (ret < 0)
+		return ret;
+
+	*val = temp & 0x0FFF;
```
Could this masking operation be a little more documented ? Maybe a #define with proper name or comment.

layman-n-ish replied:
Yes, makes sense. Addressed in v4.

buha commented on drivers/iio/gyro/adxrs290.c:
```diff
@@ -5,23 +5,344 @@
+	best_freq_idx = 0;
+	best_diff = UINT_MAX;
+	for (i = 0; i < n; i++) {
+		diff = ((freq_tbl[i][0] >= freq) ? freq_tbl[i][0] - freq
```
why not abs() ?

layman-n-ish replied:
I don't remember why I chose this way; using `abs()` is obviously much "cleaner". Acknowledged in v4.

buha commented on drivers/iio/gyro/adxrs290.c:
```diff
@@ -5,23 +5,344 @@
+	struct adxrs290_state *st = iio_priv(indio_dev);
+	u8 val;
+
+	val = ((hpf_idx << 4) & ADXRS290_HPF_MASK) |
```
There seems to be a strong preference in linux to use FIELD_GET/FIELD_PREP so maybe it would be good to see how it's typically done and try to comply.

layman-n-ish replied:
Sorry, wasn't aware of those macros. They seem very convenient. Incorporated those in v4. Thanks for the review! :D

layman-n-ish commented:
Changelog v4:

- equip `abs()` when we find the closest {LPF, HPF} 3db frequency match from a list of available {LPF, HPF} 3db frequencies. Include `<linux/kernel.h>` for the same.
- add a comment when we extract the lower 12 bits temperature data to reduce ambiguity around it.
- utilize the convenient FIELD_{PREP, GET} macros when reading/setting filter regs. Include `<linux/bitfield.h>` for the same. Change the function signature of `adxrs290_get_3db_freq()` appropriately such that the mask for `FIELD_GET()` is a compilation time constant, as documented [here](https://elixir.bootlin.com/linux/v4.9/source/include/linux/bitfield.h#L25). This (amongst many other ways), also, pushes us to remove the need of making two consequent calls to `adxrs290_get_3db_freq()` - once to get the LPF frequency and then to get the HPF frequency.
